src/Features/Core/Portable/EditAndContinue/EditAndContinueDiagnosticUpdateSource.cs

```diff
@@ -64,56 +64,53 @@ public void ReportDiagnostics(Workspace workspace, Solution solution, ProjectId?
                 return;
             }
 
-            using var _1 = ArrayBuilder<DiagnosticData>.GetInstance(out var documentDiagnosticData);
-            using var _2 = ArrayBuilder<DiagnosticData>.GetInstance(out var nonDocumentDiagnosticData);
-            var options = solution.Options;
-            var project = (projectId != null) ? solution.GetProject(projectId) : null;
+            var documentDiagnostics = diagnostics.WhereAsArray(d => d.DocumentId != null);
+            var projectDiagnostics = diagnostics.WhereAsArray(d => d.DocumentId == null && d.ProjectId != null);
+            var solutionDiagnostics = diagnostics.WhereAsArray(d => d.DocumentId == null && d.ProjectId == null);
 
-            foreach (var diagnostic in diagnostics)
+            if (documentDiagnostics.Length > 0)
             {
-                var document = solution.GetDocument(diagnostic.Location.SourceTree);
-
-                if (document != null)
-                {
-                    documentDiagnosticData.Add(DiagnosticData.Create(diagnostic, document));
-                }
-                else if (project != null)
+                foreach (var (documentId, diagnosticData) in documentDiagnostics.ToDictionary(data => data.DocumentId!))
                 {
-                    nonDocumentDiagnosticData.Add(DiagnosticData.Create(diagnostic, project));
-                }
-                else
-                {
-                    nonDocumentDiagnosticData.Add(DiagnosticData.Create(diagnostic, options));
+                    var diagnosticGroupId = (this, documentId, documentId.ProjectId);
```

Comment:
is ProjectId reundant here given docId already as the second tuple value?

Reply:
Yes.